drivers/temperature/adt7420/iio_adt7420.c

```diff
@@ -0,0 +1,530 @@
+ *
+ * @return ret - Result of the remove procedure.
+*******************************************************************************/
+int adt7420_iio_remove(struct adt7420_iio_dev *desc)
+{
+	int ret;
+
+	ret = adt7420_remove(desc->adt7420_dev);
+	if (ret)
+		return ret;
+
+	free(desc);
+	return 0;
+}
+
+/**
+ * @brief adt7420 reg read wrapper
+ *
+ * @param dev - The iio device structure.
+ * @param reg - Register address
+ * @param readval - Register value
+ *
+ * @return 0 in case of success, errno errors otherwise
+ */
+static int adt7420_iio_reg_read(struct adt7420_iio_dev *dev, uint32_t reg,
+				uint32_t *readval)
+{
+	if (reg > __UINT8_MAX__)
+		return -EINVAL;
+	return adt7420_reg_read(dev->adt7420_dev, reg, readval);
+}
+
+/**
+ * @brief adt7420 reg write wrapper
+ *
+ * @param dev - The iio device structure.
+ * @param reg - Register address
+ * @param writeval - THe value to be written in register
+ *
+ * @return 0 in case of success, errno errors otherwise
+ */
+static int adt7420_iio_reg_write(struct adt7420_iio_dev *dev, uint32_t reg,
+				 uint32_t writeval)
+{
+	if (!dev)
+		return -ENODEV;
+
+	if (writeval > __UINT16_MAX__)
+		return -EINVAL;
+
+	return adt7420_reg_write(dev->adt7420_dev, reg, writeval);
+
+}
+
+/**
+ * @brief Handles the read request for temp attribute.
+ *
+ * @param dev     - The iio device structure.
+ * @param buf	  - Command buffer to be filled with requested data.
+ * @param len     - Length of the received command buffer in bytes.
+ * @param channel - Command channel info.
+ * @param priv    - Command attribute id.
+ *
+ * @return ret    - 0 in case of success, errno errors otherwise
+*/
+static int adt7420_iio_read_temp(void *dev, char *buf, uint32_t len,
+				 const struct iio_ch_info *channel, intptr_t *priv)
+{
+	struct adt7420_iio_dev *iio_adt7420;
+	struct adt7420_dev *adt7420;
+	if (!dev)
+		return -ENODEV;
+
+	int32_t val;
+	iio_adt7420 = (struct adt7420_iio_dev *) dev;
+	adt7420 = iio_adt7420->adt7420_dev;
+
+	switch (channel->address) {
+	case 0:
+		val = adt7420_get_temperature(adt7420);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return iio_format_value(buf, len, IIO_VAL_INT, 1, &val);
+}
+
+/**
+ * @brief Handles the read request for max attribute.
+ *
+ * @param dev     - The iio device structure.
+ * @param buf	  - Command buffer to be filled with requested data.
+ * @param len     - Length of the received command buffer in bytes.
+ * @param channel - Command channel info.
+ * @param priv    - Command attribute id.
+ *
+ * @return ret    - 0 in case of success, errno errors otherwise
+*/
+static int adt7420_iio_read_max(void *dev, char *buf, uint32_t len,
+				const struct iio_ch_info *channel, intptr_t *priv)
+{
+	struct adt7420_iio_dev *iio_adt7420;
+	struct adt7420_dev *adt7420;
+
+	uint16_t temp = 0;
+	int32_t temp_c = 0;
+	iio_adt7420 = (struct adt7420_iio_dev *) dev;
+	adt7420 = iio_adt7420->adt7420_dev;
+
+	switch (channel->address) {
+	case 0:
+		if (adt7420_is_spi(adt7420))
```

Comment:
Shouldn't this interface logic be handled in the device driver itself ? 
IIO part of the driver seems like an odd place to do this to me.
Obviously this implies some small changes to the driver itself.

Reply:
ok, this family is a bit odd, it has different way of addressing the registers on spi and on i2c, so i guess this is fine